src/inventory_ui.cpp

```diff
@@ -56,48 +129,260 @@ const item &inventory_entry::get_item() const {
+}
+
+std::string inventory_selector_preset::get_cell_text( const inventory_entry &entry,
+        size_t cell_index ) const
+{
+    if( cell_index >= cells.size() ) {
+        debugmsg( "Invalid cell index %d.", cell_index );
+        return bug_text;
+    }
+    if( entry.is_null() ) {
+        return std::string();
+    } else if( entry.is_item() ) {
+        return cells[cell_index].second( entry );
+    } else if( cell_index != 0 ) {
+        return cells[cell_index].first;
+    } else {
+        return entry.get_category_ptr()->name;
+    }
+}
+
+nc_color inventory_selector_preset::get_cell_color( const inventory_entry &entry, size_t ) const
+{
+    return entry.is_item() ? get_color( entry.get_location() ) : c_magenta;
+}
+
+size_t inventory_selector_preset::get_cell_width( const inventory_entry &entry,
+        size_t cell_index ) const
+{
+    return utf8_width( get_cell_text( entry, cell_index ), true );
+}
+
+void inventory_selector_preset::append_cell( const item_text_func &func, const std::string &title )
+{
+    append_cell( [ func ]( const inventory_entry & entry ) { // Don't pass by reference
+        return func( entry.get_item() );
+    }, title );
+}
+
+void inventory_selector_preset::append_cell( const item_location_text_func &func,
+        const std::string &title )
+{
+    append_cell( [ func ]( const inventory_entry & entry ) { // Don't pass by reference
+        return func( entry.get_location() );
+    }, title );
+}
+
+void inventory_selector_preset::append_cell( const entry_text_func &func, const std::string &title )
+{
+    const auto iter = std::find_if( cells.begin(), cells.end(), [ &title ]( const cell_pair & cell ) {
+        return cell.first == title;
+    } );
+    if( iter != cells.end() ) {
+        debugmsg( "Tried to append a duplicate cell \"%s\": ignored.", title.c_str() );
+        return;
+    }
+    cells.emplace_back( title, func );
+}
+
+inventory_column::inventory_column( const inventory_selector_preset &preset ) :
+    preset( preset )
+{
+    cell_widths.resize( preset.get_cells_count(), 0 );
+}
+
+bool inventory_column::activatable() const
+{
+    return std::any_of( entries.begin(), entries.end(), []( const inventory_entry & entry ) {
+        return entry.is_selectable();
+    } );
+}
+
+inventory_entry *inventory_column::find_by_invlet( long invlet )
+{
+    for( auto &entry : entries ) {
         if( entry.is_item() && ( entry.get_invlet() == invlet ) ) {
-            return const_cast<inventory_entry *>( &entry );
+            return &entry;
         }
     }
     return nullptr;
 }
 
-size_t inventory_column::get_width() const
+size_t inventory_column::get_entry_indent( const inventory_entry &entry ) const
 {
-    size_t res = 0;
-    for( const auto &entry : entries ) {
-        res = std::max( res, get_entry_width( entry ) );
+    if( !entry.is_item() ) {
+        return 0;
+    }
+    size_t res = 2;
+    if( OPTIONS["ITEM_SYMBOLS"] ) {
```

Comment:
> I don't care about astyle here

Except for having to work around astyle bugs there is no good reason not to care